File: frontend_flutter/lib/services/api_client.dart
import 'package:dio/dio.dart';
import '../core/database/database_helper.dart';

class ApiClient {
  late final Dio dio;

  ApiClient() {
    dio = Dio(
      BaseOptions(
        connectTimeout: const Duration(seconds: 15),
        receiveTimeout: const Duration(seconds: 15),
        contentType: Headers.jsonContentType,
      ),
    );

    dio.interceptors.add(
      InterceptorsWrapper(
        onRequest: (options, handler) {
          // Dynamically fetch base URL from Hive settings cache
          final settings = DatabaseHelper.getSettings();
          options.baseUrl = settings.apiUrl;

          // Append access token if cached
          final token = DatabaseHelper.getAccessToken();
          if (token != null) {
            options.headers['Authorization'] = 'Bearer $token';
          }
          return handler.next(options);
        },
        onError: (DioException error, handler) async {
          // Handle access token expiration and refresh automatically
          if (error.response?.statusCode == 401 && 
              error.response?.data?['code'] == 'TOKEN_EXPIRED') {
            
            final success = await _rotateTokens();
            if (success) {
              // Retry original request with new access token
              final requestOptions = error.requestOptions;
              final newToken = DatabaseHelper.getAccessToken();
              requestOptions.headers['Authorization'] = 'Bearer $newToken';
              
              // Set dynamically updated base url
              final settings = DatabaseHelper.getSettings();
              requestOptions.baseUrl = settings.apiUrl;

              try {
                final response = await dio.fetch(requestOptions);
                return handler.resolve(response);
              } on DioException catch (retryErr) {
                return handler.next(retryErr);
              }
            }
          }
          return handler.next(error);
        },
      ),
    );
  }

  /// Lock client, query backend to refresh JWT access keys, and save.
  Future<bool> _rotateTokens() async {
    final refreshKey = DatabaseHelper.getRefreshToken();
    if (refreshKey == null) return false;

    final settings = DatabaseHelper.getSettings();
    final url = '${settings.apiUrl}/api/auth/refresh-token';

    try {
      // Use clean Dio instance to bypass interceptors loop
      final rotateClient = Dio();
      final response = await rotateClient.post(
        url,
        data: {'refreshToken': refreshKey},
      );

      if (response.statusCode == 200 && response.data['success'] == true) {
        final accessToken = response.data['tokens']['accessToken'] as String;
        final newRefresh = response.data['tokens']['refreshToken'] as String;

        await DatabaseHelper.saveTokens(accessToken, newRefresh);
        return true;
      }
    } catch (e) {
      // Refresh token is expired or revoked. Wipe local tokens to force login redirection.
      await DatabaseHelper.clearAuthData();
    }
    return false;
  }
}

File: frontend/src/services/apiClient.js
import axios from 'axios';
import { DatabaseHelper } from './storage';

const apiClient = axios.create({
  timeout: 15000,
  headers: {
    'Content-Type': 'application/json',
  },
});

// Request Interceptor: Attach dynamic Base URL and Access Token
apiClient.interceptors.request.use(
  (config) => {
    const settings = DatabaseHelper.getSettings();
    config.baseURL = settings.apiUrl;

    const token = DatabaseHelper.getAccessToken();
    if (token) {
      config.headers['Authorization'] = `Bearer ${token}`;
    }
    return config;
  },
  (error) => {
    return Promise.reject(error);
  }
);

// Response Interceptor: Auto token rotation on expired accessToken
apiClient.interceptors.response.use(
  (response) => response,
  async (error) => {
    const originalRequest = error.config;

    // Check if error is 401 and not a retry attempt
    if (
      error.response &&
      error.response.status === 401 &&
      error.response.data &&
      error.response.data.code === 'TOKEN_EXPIRED' &&
      !originalRequest._retry
    ) {
      originalRequest._retry = true;

      try {
        const success = await rotateTokens();
        if (success) {
          const newToken = DatabaseHelper.getAccessToken();
          originalRequest.headers['Authorization'] = `Bearer ${newToken}`;
          
          const settings = DatabaseHelper.getSettings();
          originalRequest.baseURL = settings.apiUrl;

          return apiClient(originalRequest);
        }
      } catch (retryErr) {
        return Promise.reject(retryErr);
      }
    }

    return Promise.reject(error);
  }
);

/**
 * Perform token rotation via refresh endpoint
 */
async function rotateTokens() {
  const refreshToken = DatabaseHelper.getRefreshToken();
  if (!refreshToken) return false;

  const settings = DatabaseHelper.getSettings();
  const url = `${settings.apiUrl}/api/auth/refresh-token`;

  try {
    const response = await axios.post(url, { refreshToken });
    if (response.status === 200 && response.data.success === true) {
      const accessToken = response.data.tokens.accessToken;
      const newRefresh = response.data.tokens.refreshToken;

      DatabaseHelper.saveTokens(accessToken, newRefresh);
      return true;
    }
  } catch (e) {
    console.error('[apiClient] Refresh token rotation failed', e);
    DatabaseHelper.clearAuthData();
  }
  return false;
}

export default apiClient;
